File: ios/FlowKit/Sources/FlowKit/Connection/UploadClient.swift
// `POST /upload` (shared/protocol/remotly-protocol.md §2): a photo from the phone becomes a file on the host
// that the program in a pane reads by path. Same origin, TLS trust and device token as the WebSocket.
import Foundation

public struct UploadResult: Decodable, Hashable, Sendable {
    /// Absolute path of the stored file on the host.
    public var path: String
    public var bytes: Int
}

public enum UploadError: Error, Hashable, Sendable, LocalizedError {
    case auth
    case forbidden
    case tooLarge
    case unsupportedType
    case server(status: Int)
    case transport(String)

    public var errorDescription: String? {
        switch self {
        case .auth: return "The bridge no longer accepts this device. Pair again."
        case .forbidden: return "The bridge refused this device (not on the tailnet?)."
        case .tooLarge: return "The photo is too large for the bridge."
        case .unsupportedType: return "The bridge accepts JPEG and PNG only."
        case .server(let status): return "Upload failed (HTTP \(status))."
        case .transport(let message): return message
        }
    }
}

public struct UploadClient: Sendable {
    public let host: PairedHost

    public init(host: PairedHost) {
        self.host = host
    }

    public func upload(_ data: Data, contentType: String = "image/jpeg") async throws -> UploadResult {
        var request = URLRequest(url: host.restBaseURL.appending(path: "upload"))
        request.httpMethod = "POST"
        request.setValue(contentType, forHTTPHeaderField: "Content-Type")
        request.setValue("Bearer \(host.token)", forHTTPHeaderField: "Authorization")
        request.timeoutInterval = 60
        let config = URLSessionConfiguration.ephemeral
        config.waitsForConnectivity = false
        let session = URLSession(configuration: config, delegate: PinningSessionDelegate(fingerprint: host.fingerprint), delegateQueue: nil)
        defer { session.finishTasksAndInvalidate() }
        let body: Data
        let response: URLResponse
        do {
            (body, response) = try await session.upload(for: request, from: data)
        } catch {
            throw UploadError.transport(error.localizedDescription)
        }
        switch (response as? HTTPURLResponse)?.statusCode ?? 0 {
        case 200:
            do { return try JSONDecoder().decode(UploadResult.self, from: body) } catch { throw UploadError.server(status: 200) }
        case 401: throw UploadError.auth
        case 403: throw UploadError.forbidden
        case 413: throw UploadError.tooLarge
        case 415: throw UploadError.unsupportedType
        case let status: throw UploadError.server(status: status)
        }
    }
}

File: android/core/src/main/kotlin/com/inferenceaftermath/remotly/core/connection/UploadClient.kt
// POST /upload (protocol §2): a photo from the phone becomes a file on the host that the program in a pane
// reads by path. Same origin, TLS trust and device token as the WebSocket.
package com.inferenceaftermath.remotly.core.connection

import com.inferenceaftermath.remotly.core.pairing.FlowTls
import com.inferenceaftermath.remotly.core.pairing.QrPayload
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.json.Json
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.util.concurrent.TimeUnit

/** @property path absolute path of the stored file on the host. */
@Serializable
data class UploadResult(val path: String, val bytes: Long = 0)

@Serializable
private data class UploadErrorBody(val error: String = "")

class UploadException(val code: String, val httpStatus: Int) : Exception(describe(code, httpStatus)) {
    companion object {
        fun describe(code: String, status: Int): String = when (code) {
            "auth" -> "The bridge no longer accepts this device; pair again"
            "forbidden" -> "This device is not on the host's tailnet"
            "too_large" -> "The photo is too large for the bridge"
            "unsupported_type" -> "The bridge accepts JPEG and PNG only"
            "network" -> "Could not reach the host"
            else -> "Upload failed (HTTP $status)"
        }
    }
}

class UploadClient(
    private val host: HostConfig,
    private val clientFor: (String?) -> OkHttpClient = { fp ->
        FlowTls.client(fp, OkHttpClient.Builder().connectTimeout(10, TimeUnit.SECONDS).writeTimeout(60, TimeUnit.SECONDS).readTimeout(30, TimeUnit.SECONDS))
    },
) {
    private val json = Json { ignoreUnknownKeys = true }

    suspend fun upload(bytes: ByteArray, contentType: String = "image/jpeg"): UploadResult = withContext(Dispatchers.IO) {
        val request = Request.Builder()
            .url(QrPayload.httpsOrigin(host.url) + "/upload")
            .header("Authorization", "Bearer ${host.token}")
            .post(bytes.toRequestBody(contentType.toMediaType()))
            .build()
        val response = try {
            clientFor(host.fingerprint).newCall(request).execute()
        } catch (e: IOException) {
            throw UploadException("network", 0).initCause(e)
        }
        response.use { r ->
            val text = r.body?.string() ?: ""
            if (r.code == 200) {
                try {
                    json.decodeFromString(UploadResult.serializer(), text)
                } catch (e: Exception) {
                    throw UploadException("bad_response", r.code)
                }
            } else {
                val code = try { json.decodeFromString(UploadErrorBody.serializer(), text).error } catch (e: Exception) { "" }
                throw UploadException(code.ifEmpty { "http_${r.code}" }, r.code)
            }
        }
    }
}
